lib/mantav2-migrate/migrator.js

```diff
@@ -0,0 +1,746 @@
+    for (let shardIdx = 0; shardIdx < opts.indexShards.length; shardIdx++) {
+        let shard = opts.indexShards[shardIdx];
+
+        let content;
+        let records;
+        try {
+            content = fs.readFileSync(shard.discoveryFile);
+        } catch (readErr) {
+            cb(readErr);
+            return;
+        }
+        try {
+            records = zlib
+                .gunzipSync(content)
+                .toString('utf8')
+                .trim()
+                .split(/\n/g)
+                .map(line => JSON.parse(line));
+        } catch (parseErr) {
+            cb(new VError(parseErr, 'could not load %s', shard.discoveryFile));
+            return;
+        }
+        log.debug(
+            {discoveryFile: shard.discoveryFile, numRecords: records.length},
+            'loaded discovery file'
+        );
+
+        for (let i = 0; i < records.length; i++) {
+            let obj = records[i];
+            let newUuid = uuidv4();
+
+            if (obj.size === 0) {
+                numZeroByteObjs++;
+            } else {
+                for (let s = 0; s < obj.storageIds.length; s++) {
+                    let storId = obj.storageIds[s];
+                    if (!storCmdsFromStorId.hasOwnProperty(storId)) {
+                        storCmdsFromStorId[storId] = [];
+                    }
+
+                    // Add a `ln <old> <new>` for each storage zone that has
+                    // this object. To make this re-runnable, we skip that if
+                    // the two exist and refer to the same file (hardlink or
+                    // symlink).
+                    //
+                    //      if [[ ! $old -ef $new ]]; then
+                    //          ln $old $new
+                    //      fi
+                    let oldPath = `/manta/${obj.creatorId}/${obj.objectId}`;
+                    let newPath = `/manta/${obj.creatorId}/${newUuid}`;
+                    storCmdsFromStorId[storId].push(
+                        `if [[ ! ${oldPath} -ef ${newPath} ]]; then`
+                    );
+                    storCmdsFromStorId[storId].push(
+                        `    ln ${oldPath} ${newPath}`
+                    );
+                    storCmdsFromStorId[storId].push(`fi`);
+                }
+            }
+
+            // Our metadata command cannot handle some special chars in
+            // the object key.
+            if (obj.key.indexOf('"') !== -1 || obj.key.indexOf('\\') !== -1) {
+                cb(
+                    new VError(
+                        'object key cannot have `"` or `\\` characters: %s',
+                        obj.key
+                    )
+                );
+                return;
+            }
+
+            if (!metadataCmdsFromShard.hasOwnProperty(shard.host)) {
+                metadataCmdsFromShard[shard.host] = [];
+            }
+
+            // Add a command to update the metadata entry to the new object
```

Comment:
I'm curious - for commands that don't actually have to touch the stored objects on disk, is there any way we could invoke a single command to process all of the metadata updates in bulk? I'm not sure how many snaplinks we have in production, but if the answer is "lots", then cutting down on the amount of process-execution might be a win.

Reply:
Point taken. Per https://github.com/joyent/moray/blob/master/docs/index.md#batch there *is* a Moray "batch". However for the following reasons I think I'd prefer to stick with the current putobject-for-each-snaplink:

- There isn't a ready available CLI for the moray "batch" RPC: https://github.com/joyent/node-moray/tree/master/bin   We would have to cobble something and that would add some complexity here.
- I don't think we expect there to be zillions of snaplinks. There weren't crazy amounts in eng staging when josh was looking. I could be wrong in JPC. 
- We expect the time to run the sherlock scripts to vastly dominate the full runtime here, so I don't believe optimizing this step will be a huge overall win.